libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/location/LocationValidator.java

```diff
@@ -0,0 +1,106 @@
+package com.mapbox.services.android.navigation.v5.location;
+
+import android.location.Location;
+
+import com.mapbox.geojson.Point;
+import com.mapbox.turf.TurfConstants;
+import com.mapbox.turf.TurfMeasurement;
+
+import java.util.concurrent.TimeUnit;
+
+import timber.log.Timber;
+
+public class LocationValidator {
+
+  private static final int TIME_THRESHOLD = 5000; // 5 seconds
+  private static final int ACCURACY_THRESHOLD = 10; // percent
+  private static final int VELOCITY_THRESHOLD = 200; // milliseconds
+
+  private static LocationValidator instance;
+  private Location previousValidLocation;
+  private boolean loggingEnabled;
+
+  /**
+   * Primary access method (using singleton pattern)
+   *
+   * @return LocationValidator
+   */
+  public static synchronized LocationValidator getInstance() {
+    if (instance == null) {
+      instance = new LocationValidator();
+    }
+
+    return instance;
+  }
+
+  public boolean isValidUpdate(Location location) {
+    // First update
+    if (previousValidLocation == null) {
+      previousValidLocation = location;
+      return true;
+    }
+
+    float currentAccuracy = location.getAccuracy();
+    float previousAccuracy = previousValidLocation.getAccuracy();
+    float accuracyDifference = Math.abs(previousAccuracy - currentAccuracy);
+
+    log("Accuracy - Current: %s, Previous: %s, Difference: %s",
+      currentAccuracy, previousAccuracy, accuracyDifference);
+
+    boolean currentAccuracyWorse = currentAccuracy > previousAccuracy;
+    boolean hasSameProvider = previousValidLocation.getProvider().equals(location.getProvider());
+    boolean lessThanPercentThreshold = (accuracyDifference <= (previousAccuracy / ACCURACY_THRESHOLD));
+
+    // New location update is acceptable, even with worse accuracy, if it is from
+    // the same provider and is no more than 10% worse
+    boolean worseAccuracyAcceptable = currentAccuracyWorse && hasSameProvider && lessThanPercentThreshold;
```

Comment:
maybe reword this to `accuracyAcceptable`? because it doesn't really make sense if `currentAccuracyWorse` is false.

Reply:
Yeah good point, naming was bad here